controllers/upgrade_handlers.go

```diff
@@ -54,9 +54,91 @@ func (b BackupValidationError) Error() string {
 
 // handleUpgrade orchestrate main upgrade steps and update status as needed
 func (r *ImageBasedUpgradeReconciler) handleUpgrade(ctx context.Context, ibu *lcav1alpha1.ImageBasedUpgrade) (ctrl.Result, error) {
-
 	r.Log.Info("Starting handleUpgrade")
 
+	currentStaterootName, _ := r.RPMOstreeClient.GetCurrentStaterootName()
+	r.Log.Info("stateroots", "current stateroot:", currentStaterootName, "desired stateroot", getDesiredStaterootName(ibu))
+	if getDesiredStaterootName(ibu) != currentStaterootName {
+		if r.completedPrePivotOnce {
+			//todo: abort handler? e.g delete desired stateroot?
+			msg := "pre-pivot steps are allowed to run only once"
+			r.Log.Error(fmt.Errorf(msg), "")
+			upgradeFailedStatus(ibu, msg)
+			return doNotRequeue(), nil
+		}
+
+		r.Log.Info("Starting pre pivot steps and will pivot to new stateroot with a reboot")
+		ctrlResult, err := r.prePivot(ctx, ibu)
+		if err != nil {
+			//todo: abort handler? e.g delete desired stateroot
+			return requeueWithError(err)
+		}
+
+		// prePivot requested a requeue
+		if !ctrlResult.IsZero() {
+			return ctrlResult, nil
+		}
+
+		// set this to true before a reboot
+		r.completedPrePivotOnce = true
```

Comment:
This is just to prevent redoing prePivot before reboot/operator restart?

Reply:
Yeah that was the intention...but it wasnt too necessary in first place. But I made some adjustments to make sure prepivot run only ones. PTAL?

Reply:
Looks good.  So prePivot could still take multiple attempts. Reboot is only tried once with the sleep.